pkg/controllers/jobset_controller_test.go

```diff
@@ -263,3 +265,305 @@ func TestSetExclusiveAffinities(t *testing.T) {
+				},
+			},
+			want: []*batchv1.Job{
+				makeJob(&expectedJobArgs{
+					jobName:  "test-jobset-replicated-job-1",
+					ns:       ns,
+					replicas: 2,
+					jobIdx:   1}),
+			},
+		},
+		{
+			name: "multiple replicated jobs",
+			js: testutils.MakeJobSet(jobSetName, ns).
+				ReplicatedJob(testutils.MakeReplicatedJob("replicated-job-A").
+					Job(testutils.MakeJobTemplate(jobName, ns).Obj()).
+					Replicas(1).
+					Obj()).
+				ReplicatedJob(testutils.MakeReplicatedJob("replicated-job-B").
+					Job(testutils.MakeJobTemplate(jobName, ns).Obj()).
+					Replicas(2).
+					Obj()).
+				Obj(),
+			ownedJobs: &childJobs{
+				active: []*batchv1.Job{
+					makeJob(&expectedJobArgs{
+						jobName:  "test-jobset-replicated-job-B-0",
+						ns:       ns,
+						replicas: 2,
+						jobIdx:   0}),
+				},
+			},
+			want: []*batchv1.Job{
+				makeJob(&expectedJobArgs{
+					jobName:  "test-jobset-replicated-job-A-0",
+					ns:       ns,
+					replicas: 1,
+					jobIdx:   0}),
+				makeJob(&expectedJobArgs{
+					jobName:  "test-jobset-replicated-job-B-1",
+					ns:       ns,
+					replicas: 2,
+					jobIdx:   1}),
+			},
+		},
+		{
+			name: "exclusive affinities",
+			js: testutils.MakeJobSet(jobSetName, ns).
+				ReplicatedJob(testutils.MakeReplicatedJob("replicated-job").
+					Job(testutils.MakeJobTemplate(jobName, ns).Obj()).
+					Replicas(1).
+					Exclusive(exclusive).
+					Obj()).
+				Obj(),
+			ownedJobs: &childJobs{},
+			want: []*batchv1.Job{
+				makeJob(&expectedJobArgs{
+					jobName:   "test-jobset-replicated-job-0",
+					ns:        ns,
+					replicas:  1,
+					exclusive: exclusive,
+					jobIdx:    0}),
+			},
+		},
+		{
+			name: "pod dns hostnames enabled",
+			js: testutils.MakeJobSet(jobSetName, ns).
+				ReplicatedJob(testutils.MakeReplicatedJob("replicated-job").
+					Job(testutils.MakeJobTemplate(jobName, ns).Obj()).
+					Replicas(1).
+					EnableDNSHostnames(true).
+					Obj()).
+				Obj(),
+			ownedJobs: &childJobs{},
+			want: []*batchv1.Job{
+				makeJob(&expectedJobArgs{
+					jobName:   "test-jobset-replicated-job-0",
+					ns:        ns,
+					replicas:  1,
+					subdomain: "test-jobset-replicated-job-0",
+					jobIdx:    0}),
+			},
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			var got []*batchv1.Job
+			for _, rjob := range tc.js.Spec.Jobs {
+				jobs, err := constructJobsFromTemplate(tc.js, &rjob, tc.ownedJobs)
+				if err != nil {
+					t.Errorf("constructJobsFromTemplate() error = %v", err)
+					return
+				}
+				got = append(got, jobs...)
+			}
+
+			if diff := cmp.Diff(tc.want, got); diff != "" {
+				t.Errorf("constructJobsFromTemplate() mismatch (-want +got):\n%s", diff)
+			}
+		})
+	}
+}
+
+type expectedJobArgs struct {
+	jobName   string
+	ns        string
+	subdomain string
+	replicas  int
+	jobIdx    int
+	restarts  int
+	exclusive *jobset.Exclusive
+}
+
+func makeJob(args *expectedJobArgs) *batchv1.Job {
```

Comment:
I would make this return a job wrapper and remove the subdomain and exclusive args since each are used in a single test.

For example, the subdomain test will look like: `makeJob().Subdomain("test-jobset-replicated-job-0")`; same with the exclusive test.

Reply:
Makes sense, done.